physics/forkable_body.hpp

```diff
@@ -0,0 +1,240 @@
+  auto& child_forkable = child_it->second;
+  child_forkable.parent_ = that();
+  child_forkable.position_in_parent_children_ = child_it;
+  child_forkable.position_in_parent_timeline_ = timeline_it;
+
+  // Copy the tail of the trajectory in the child object.
+  if (timeline_it != timeline_end()) {
+    child_forkable.timeline_insert(++timeline_it, timeline_end());
+  }
+  return &child_forkable;
+}
+
+template<typename Tr4jectory>
+void Forkable<Tr4jectory>::DeleteFork(
+    not_null<Tr4jectory**> const trajectory) {
+  CHECK_NOTNULL(*trajectory);
+  Instant const* const fork_time = (*trajectory)->ForkTime();
+  CHECK_NOTNULL(fork_time);
+  // Find the position of |*forkable| among our children and remove it.
+  auto const range = children_.equal_range(*fork_time);
+  for (auto it = range.first; it != range.second; ++it) {
+    if (&it->second == *trajectory) {
+      children_.erase(it);
+      *trajectory = nullptr;
+      return;
+    }
+  }
+  LOG(FATAL) << "argument is not a child of this trajectory";
+}
+
+template<typename Tr4jectory>
+bool Forkable<Tr4jectory>::is_root() const {
+  return parent_ == nullptr;
+}
+
+template<typename Tr4jectory>
+not_null<Tr4jectory const*>
+Forkable<Tr4jectory>::root() const {
+  Tr4jectory const* ancestor = that();
+  while (ancestor->parent_ != nullptr) {
+    ancestor = ancestor->parent_;
+  }
+  return ancestor;
+}
+
+template<typename Tr4jectory>
+not_null<Tr4jectory*>
+Forkable<Tr4jectory>::root() {
+  Tr4jectory* ancestor = that();
+  while (ancestor->parent_ != nullptr) {
+    ancestor = ancestor->parent_;
+  }
+  return ancestor;
+}
+
+template<typename Tr4jectory>
+Instant const* Forkable<Tr4jectory>::ForkTime() const {
+  if (is_root()) {
+    return nullptr;
+  } else {
+    return &position_in_parent_children_->first;
+  }
+}
+
+template<typename Tr4jectory>
+bool Forkable<Tr4jectory>::Iterator::operator==(
+    Iterator const & right) const {
+  bool const this_at_end = at_end();
+  bool const right_at_end = right.at_end();
+  if (this_at_end != right_at_end) {
+    return false;
+  } else if (this_at_end) {
+    return true;
+  } else {
+    return ancestry_ == right.ancestry_ && current_ == right.current_;
+  }
+}
+
+template<typename Tr4jectory>
+bool Forkable<Tr4jectory>::Iterator::operator!=(
+    Iterator const & right) const {
+  return !(*this == right);
+}
+
+template<typename Tr4jectory>
+typename Forkable<Tr4jectory>::Iterator&
+Forkable<Tr4jectory>::Iterator::operator++() {
+  CHECK(!ancestry_.empty());
+  CHECK(!at_end());
+
+  // Check if there is a next child in the ancestry.
+  auto ancestry_it = ancestry_.begin();
+  if (++ancestry_it != ancestry_.end()) {
+    // There is a next child.  See if we reached its fork time.
+    Instant const& current_time = ForkableTraits<Tr4jectory>::time(current_);
+    not_null<Tr4jectory const*> child = *ancestry_it;
+    Instant child_fork_time = child->position_in_parent_children_->first;
+    if (current_time == child_fork_time) {
```

Comment:
No, I think it works, added test IteratorIncrementMultipleForksSuccess to exercise the case that you were concerned about.